Rewrite (#2)
mihaibudiu/sigmod-rh: recursion.tex

\section{Recursive queries}\label{sec:recursion}

Recursive queries are very useful in a many applications.  For
example, graph algorithms such as graph reachability or transitive
closure are naturally expressed using recursive queries.  We introduce
two simple \dbsp stream operators that are used for expressing
recursive query evaluation.  These operators allow us to build
circuits implementing looping constructs, which are used to iterate
computations until a fixed-point is reached.

\subsection{Creating and destroying streams}

%\begin{definition}\label{def:zae}
%We say that a stream $s \in \stream{A}$ is \defined{zero almost-everywhere} if it has a finite
%number of non-zero values, i.e., there exists a time $t_0 \in \N$
%s.t. $\forall t \geq t_0 . s[t] = 0$.
%\noindent Denote the set of streams that are zero almost everywhere
%by $\streamf{A} \subset \stream{A}$.
%\end{definition}

% \paragraph{Stream introduction}

> The \defined{delta function} (named from the Dirac delta function)

Reviewer: not sure what "named from" means. Can we say that it _is_ the Dirac function?

\end{tikzpicture}
\end{center}

In this circuit $Q$ is a streaming operator, but the entire circuit
implements a scalar function, as shown by the single arrowheads.

%\begin{proposition}
%If $Q$ is time-invariant, the circuit above has the zero-preservation
%property: $\zpp{\int \circ\; Q \circ \delta_o}$.
%\end{proposition}

\subsection{Implementing recursive queries}\label{sec:datalog}

We describe the implementation of recursive queries in \dbsp.  SQL can
only express very limited recursive queries, so here we model Datalog
queries.  In general, a Datalog program defines a set of mutually
recursive relations.

> We describe here the algorithm for the simple case of a single-input,

Reviewer: ```suggestion
We describe the algorithm for the simple case of a single-input,
```

  \node[] (I) {\code{I}};
  \node[block, right of=I] (R) {$\lift R$};
  \node[right of=R, node distance=1.5cm] (O) {\code{O}};
  \node[block, below of=R, node distance=.8cm] (z) {$\zm$};
  \draw[->>] (I) -- (R);
  \draw[->>] (R) -- node(o) {} (O);
  \draw[->>] (o) |- (z);
  \draw[->>] (z) -- (R);
 \end{tikzpicture}
 \end{center}
\item ``Bracket'' the circuit as follows:

\begin{center}
\begin{tikzpicture}[auto,>=latex, node distance=.8cm]
  \node[] (Iinput) {\code{I}};
  \node[block, right of=Iinput] (ID) {$\delta_0$};
  \node[block, right of=ID] (II) {$\I$};
  \node[block, right of=II, node distance=1cm] (f) {$\lift{R}$};
  \node[block, right of=f, node distance=1.5cm] (D) {$\D$};
  \node[block, right of=D] (S) {$\int$};
  \node[right of=S] (output)  {\code{O}};
  \draw[->] (Iinput) -- (ID);
  \draw[->>] (ID) -- (II);
  \draw[->>] (II) -- (f);
  \draw[->>] (f) -- node (o) {} (D);
  \draw[->>] (D) -- (S);
  \draw[->] (S) -- (output);
  \node[block, below of=f, node distance=.8cm] (z) {$\zm$};
  \draw[->>] (o) |- (z);
  \draw[->>] (z) -- (f);
\end{tikzpicture}
\end{center}
\end{enumerate}
\end{algorithm}

Think of the feedback cycle in this circuit as a \texttt{while} loop

> that iterates until no changes are observed (in technical terms, a

Reviewer: ```suggestion
that iterates until no changes are observed (i.e., a
```
